### PR TITLE
Promises

### PR DESCRIPTION
**Pull request by marcusnaslund**

### CODE REVIEW

**Comment by davidhesselbom** [comment]
The "Accidental add" commit doesn't belong here. Clean up your history.

File: source/base/Promise.ooc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	wait: func -> T {
+		Running := true
+
+		while (Running) {
+			Time sleepMilli(10)
+			_mutex lock()
+			if (this _state != PromiseState Running) {
+			Running = false
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Indentation...

File: source/base/Promise.ooc [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		this _state = PromiseState Cancelled
+		//How to terminate thread?
+	}
+
+	_execute: override func {
+		try {
+			this _state = PromiseState Running
+
+			this _thread = Thread new(||
+			this _task()
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Indentation...

File: source/base/Promise.ooc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	wait: func ~timeout (seconds: Double) -> T {
+		_thread.wait(seconds)
+		_result
+	}
+
+	cancel: func -> Void {
+		this _state = PromiseState Cancelled
+		//How to terminate thread?
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Very good question. I don't think anybody knows...

File: source/base/Promise.ooc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+Promise: abstract class<T> {
+	_result: T
+	_state: PromiseState
+	state ::= this _state
+
+	init: func
+
+	wait: abstract func -> T { _result }
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Whenever you are using a member variable or function, use `this`: 
```ooc
wait: abstract func -> T { this _result }
```

File: source/base/Promise.ooc [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	wait: func -> T {
+		Running := true
+
+		while (Running) {
+			Time sleepMilli(10)
+			_mutex lock()
+			if (this _state != PromiseState Running) {
+			Running = false
+			}
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Don't use brackets unless you need them.

File: test/base/PromiseTest.ooc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+use ooc-math
+import math
+import os/Time
+
+PromiseTest: class extends Fixture {
+	init: func {
+		super("Promise")
+		this add("basic", func {
+			p := ThreadedPromise< Cell<Int> > new(func -> Cell<Int> {
+				Time sleepSec (1)
```
Comment (davidhesselbom) [inline test/base/PromiseTest.ooc]: `sleepSec` is a function, so `sleepSec(1)` (no space)

File: test/base/PromiseTest.ooc [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+				Cell<Int> new (42)
```
Comment (davidhesselbom) [inline test/base/PromiseTest.ooc]: `new` is also a function in ooc.

File: source/base/Promise.ooc [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	wait: abstract func -> T { _result }
+
+	wait: virtual func ~timeout (seconds: Double) -> T { _result }
+
+	cancel: virtual func -> Bool { false }
+
+	_execute: virtual func
+}
+
+ThreadedPromise : class<T> extends Promise<T> {
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Elsewhere, we've written this as `class <T>`, so do that here as well. Magic should probably catch this. @thomasfanell ?

File: source/base/Promise.ooc [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	_mutex := Mutex new()
+	_thread: Thread
+	_task: Func
+	_exception: Exception
+
+	init: func (task: Func -> T) {
+		this _task = func () -> Void {
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: I'm curious... does it work without the `()`?

File: source/base/Promise.ooc [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+			catch (e: Exception) {
+				this _exception = e
+			}
+		}
+
+		this _execute()
+	}
+
+	wait: func -> T {
+		Running := true
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Classes, not variables, should be capitalized.

File: source/base/Promise.ooc (hunk 9 above)
Comment (marcusnaslund) [inline source/base/Promise.ooc]: Yes. Will fix.

File: source/base/Promise.ooc [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+ThreadedPromise : class <T> extends Promise <T> {
+	_mutex := Mutex new()
+	_thread: Thread
+	_task: Func
+	_exception: Exception
+
+	init: func (task: Func -> T) {
+		this _task = func -> Void {
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: I don't think you need `-> Void` here, but I'm not 100% sure.

File: source/base/Promise.ooc [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+	wait: func -> T {
+		running := true
+
+		while (running) {
+			Time sleepMilli(10)
+			_mutex lock()
+			if (this _state != PromiseState Running) {
+				running = false
+			}
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Don't use brackets unless you need to.

File: source/base/Promise.ooc [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+		}
+
+		if (this _state == PromiseState Error)
+			_exception throw()
+
+		_result
+	}
+
+	wait: func ~timeout (seconds: Double) -> T {
+		_thread.wait(seconds)
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: What does the `.` do?

File: source/base/Promise.ooc [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+		//How to terminate thread?
+	}
+
+	_execute: override func {
+		try {
+			this _state = PromiseState Running
+
+			this _thread = Thread new(||
+				this _task()
+			)
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: More unnecessary brackets...

File: test/base/PromiseTest.ooc [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+use ooc-unit
+use ooc-math
+import math
+import os/Time
+
+PromiseTest: class extends Fixture {
+	init: func {
+		super("Promise")
+		this add("basic", func {
+			p := ThreadedPromise< Cell<Int> > new(func -> Cell<Int> {
```
Comment (davidhesselbom) [inline test/base/PromiseTest.ooc]: Try not to use overly short names. `promise1` shouldn't be too many letters to write.

File: source/base/Promise.ooc [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+	wait: abstract func -> T { this _result }
+
+	wait: virtual func ~timeout (seconds: Double) -> T { this _result }
+
+	cancel: virtual func -> Bool { false }
+
+	_execute: virtual func
+}
+
+ThreadedPromise : class <T> extends Promise <T> {
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: `ThreadedPromise:` (no space)

Comment (davidhesselbom) [inline source/base/Promise.ooc]: Also, I think you misunderstood me last time - do write `Promise<T>` without spaces.

**Comment by davidhesselbom** [comment]
Personally, I prefer not having this many empty lines, because all it does it makes have to scroll more. You don't need to have one after each function, for instance.

File: source/base/Promise.ooc [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+	_mutex := Mutex new()
+	init: func (=_action) {
+		super()
+		this _state = _PromiseState Unfinished
+		this _task = func {
+			this _result = this _action()
+			this _mutex lock()
+			if (this _state != _PromiseState Cancelled) {
+				this _state = _PromiseState Finished
+			}
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: You don't need these brackets.

File: source/base/Promise.ooc [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	wait: func -> Bool {
+		this _thread wait()
+		(this _state == _PromiseState Finished)
+	}
+	wait: func ~default (defaultValue: T) -> T {
+		status := this wait()
+		status ? this _result : defaultValue
+	}
+	cancel: override func -> Bool {
+		// How to end thread?
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Have you figured this out yet? If not, start the comment with `TODO: ` to make it easier to find.

**Comment by davidhesselbom** [comment]
You still need clean up history - the accidental file add doesn't belong upstream. Ask for help if you need it.

File: source/base/Promise.ooc (hunk 18 above)
Comment (marcusnaslund) [inline source/base/Promise.ooc]: I see someone else has - I will fix once that pull req. has been accepted.

**Pull request merged**